Fixes grabs not releasing on stun, headbutting

Textor44 opened the pull request:
:cl: Textor
bugfix: Going prone releases grabs.
bugfix: You can no longer keep hold of someone when you've been stunned.
bugfix: Headbutting forces you to release your grip on your victim.
/:cl:

Fixes #31177
Fixes #31347

Textor44 committed:
Fixes stun not releasing grab, fixes headbutt not releasing grab.

:cl: Textor
bugfix: People who have grabbed another person will no longer get confused by stairs and let them go.
/:cl:

Looked at how the ladder code handled climbing up ladders with grabbed people and applied the same logic.

This works both from stairs and falling, and falling with a grabbed person will cause damage to both mobs and retain the grab.

Reverts stair code, fixes stuns and headbutting.

Textor44 commented:
Adding a related bugfix into the PR.

SierraKomodo reviewed (changes requested):
Does not fully fix #31177 - You can still maintain a grab while prone from resting or other non-stun related means.

https://user-images.githubusercontent.com/11140088/165358478-ec276a35-4b53-4bb1-b9b8-518792aba6ed.mp4

The changes force you to drop a grab when stunned or in paincrit, even if still standing, which is maybe okay (Need to see how it plays out in actual combat to know for sure), but it shouldn't close an issue that's describing a different scenario.

Textor44 committed:
Falling down no longer lets you keep a grab

Adds a check to the process of the grab item that lets the grab go if the assailant is no longer upright.

Textor44 commented:
> Does not fully fix #31177 - You can still maintain a grab while prone from resting or other non-stun related means.
> 
>  dreamseeker_2022-04-26_10-29-48.mp4 
> The changes force you to drop a grab when stunned or in paincrit, even if still standing, which is maybe okay (Need to see how it plays out in actual combat to know for sure), but it shouldn't close an issue that's describing a different scenario.

Resolved.

SierraKomodo reviewed (changes requested):
It runtimes when dropping grabes while prone.

```
[23:00:18] Runtime in norm_aggressive.dm,33: Cannot read null.vars
  proc name: process effect (/datum/grab/normal/aggressive/process_effect)
  src: /datum/grab/normal/aggressive (/datum/grab/normal/aggressive)
  call stack:
  /datum/grab/normal/aggressive (/datum/grab/normal/aggressive): process effect(the grab (upper body) (/obj/item/grab/normal))
  /datum/grab/normal/aggressive (/datum/grab/normal/aggressive): process(the grab (upper body) (/obj/item/grab/normal))
  the grab (upper body) (/obj/item/grab/normal): Process()
  Floella Reade (/mob/living/carbon/human): Life()
  Floella Reade (/mob/living/carbon/human): Life()
  Floella Reade (/mob/living/carbon/human): Life()
  Mobs (/datum/controller/subsystem/mobs): fire(0, null)
  Mobs (/datum/controller/subsystem/mobs): ignite(0)
  Main (/datum/controller/master): RunQueue()
  Main (/datum/controller/master): Loop()
  Main (/datum/controller/master): StartProcessing(0)
```

Textor44 commented:
> It runtimes when dropping grabes while prone.
> 
> ```
> [23:00:18] Runtime in norm_aggressive.dm,33: Cannot read null.vars
>   proc name: process effect (/datum/grab/normal/aggressive/process_effect)
>   src: /datum/grab/normal/aggressive (/datum/grab/normal/aggressive)
>   call stack:
>   /datum/grab/normal/aggressive (/datum/grab/normal/aggressive): process effect(the grab (upper body) (/obj/item/grab/normal))
>   /datum/grab/normal/aggressive (/datum/grab/normal/aggressive): process(the grab (upper body) (/obj/item/grab/normal))
>   the grab (upper body) (/obj/item/grab/normal): Process()
>   Floella Reade (/mob/living/carbon/human): Life()
>   Floella Reade (/mob/living/carbon/human): Life()
>   Floella Reade (/mob/living/carbon/human): Life()
>   Mobs (/datum/controller/subsystem/mobs): fire(0, null)
>   Mobs (/datum/controller/subsystem/mobs): ignite(0)
>   Main (/datum/controller/master): RunQueue()
>   Main (/datum/controller/master): Loop()
>   Main (/datum/controller/master): StartProcessing(0)
> ```

Weird, I didn't encounter any runtimes when I tested them. What was your testing process? I'd like to try to reproduce it so I can see why it caused an issue.

SierraKomodo commented:
> > It runtimes when dropping grabes while prone.
> > ```
> > [23:00:18] Runtime in norm_aggressive.dm,33: Cannot read null.vars
> >   proc name: process effect (/datum/grab/normal/aggressive/process_effect)
> >   src: /datum/grab/normal/aggressive (/datum/grab/normal/aggressive)
> >   call stack:
> >   /datum/grab/normal/aggressive (/datum/grab/normal/aggressive): process effect(the grab (upper body) (/obj/item/grab/normal))
> >   /datum/grab/normal/aggressive (/datum/grab/normal/aggressive): process(the grab (upper body) (/obj/item/grab/normal))
> >   the grab (upper body) (/obj/item/grab/normal): Process()
> >   Floella Reade (/mob/living/carbon/human): Life()
> >   Floella Reade (/mob/living/carbon/human): Life()
> >   Floella Reade (/mob/living/carbon/human): Life()
> >   Mobs (/datum/controller/subsystem/mobs): fire(0, null)
> >   Mobs (/datum/controller/subsystem/mobs): ignite(0)
> >   Main (/datum/controller/master): RunQueue()
> >   Main (/datum/controller/master): Loop()
> >   Main (/datum/controller/master): StartProcessing(0)
> > ```
> 
> Weird, I didn't encounter any runtimes when I tested them. What was your testing process? I'd like to try to reproduce it so I can see why it caused an issue.

Grabbed a mob, upgraded the grab, then used rest and started moving around. I got one tile of movement while the grab was still active before the debugger caught the runtime.

Textor44 commented:
> Grabbed a mob, upgraded the grab, then used rest and started moving around. I got one tile of movement while the grab was still active before the debugger caught the runtime.

Ah, I don't think I tried to move around when I was lying down. Alright, further testing to be done.

Textor44 commented:
I'll be working on this this weekend.
